src/runtime/runtime_esp8266.go

```diff
@@ -115,3 +117,23 @@ func abort() {
 		device.Asm("waiti 0")
 	}
 }
+
+//export __sync_val_compare_and_swap_4
+func __sync_val_compare_and_swap_4(ptr *uint32, expected, desired uint32) uint32 {
```

Comment:
Huh, interesting. These functions haven't been implemented before:

```
$ build/tinygo build -o test.elf -target=esp8266 ./testdata/atomic.go
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x0): undefined reference to `__sync_fetch_and_add_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x4): undefined reference to `__sync_fetch_and_add_8'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x8): undefined reference to `__sync_lock_test_and_set_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0xc): undefined reference to `__sync_lock_test_and_set_8'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x10): undefined reference to `__sync_val_compare_and_swap_4'
/tmp/tinygo787979806/main.o:(.literal.runtime.run$1$gowrapper+0x14): undefined reference to `__sync_val_compare_and_swap_8'
/tmp/tinygo787979806/main.o:(.literal.main.testValue+0x0): undefined reference to `sync/atomic.runtime_procPin'
/tmp/tinygo787979806/main.o:(.literal.main.testValue+0x4): undefined reference to `sync/atomic.runtime_procUnpin'
error: failed to link /tmp/tinygo787979806/main: exit status 1
```

If that isn't too difficult, it would be great to have these implemented in a separate PR. Perhaps as a generalization of the functions in src/runtime/arch_cortexm.go - after all, this can be implemented the same way on all unicore processors (which are in effect all MCUs TinyGo supports as TinyGo doesn't use more than one core). I think src/runtime/baremetal.go (or even a new src/runtime/atomic.go) would be a good place for these functions.

Reply:
Yeah, I was planning on doing something like that soon.

Reply:
Moved into #2307